From eb402307c2b307fa043e7e81a989e40efaebb7e5 Mon Sep 17 00:00:00 2001
From: Steve deRosier <Steve.Derosier@lairdtech.com>
Date: Mon, 20 May 2013 10:12:18 -0700
Subject: [PATCH 2/5] Pull NAND WP# pin low ASAP

This prevents possible spurious writes/erases on the NAND flash and is as
close to the datasheet as possible. Until this point, the pin is an input
and thus prone to floating. The pin is pulled high in uboot after NAND
PMECC is enabled.

Additional problem is short glitches on this pin during ROMBoot. This fix
doesn't solve this issue, nor does it seem possible to do so.
---
 board/at91sam9x5ek/at91sam9x5ek.c |    5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/board/at91sam9x5ek/at91sam9x5ek.c b/board/at91sam9x5ek/at91sam9x5ek.c
index 03227e3..8d086ea 100644
--- a/board/at91sam9x5ek/at91sam9x5ek.c
+++ b/board/at91sam9x5ek/at91sam9x5ek.c
@@ -141,6 +141,9 @@ static void ddramc_init(void)
 #ifdef CONFIG_HW_INIT
 void hw_init(void)
 {
+	/* Make sure NAND WP# pin is low ASAP */
+	pio_set_gpio_output(AT91C_PIN_PD(10), 0);
+
 	/* Disable watchdog */
 	writel(AT91C_WDTC_WDDIS, AT91C_BASE_WDT + WDTC_MR);
 
@@ -256,7 +259,7 @@ void nandflash_hw_init(void)
 		{"NANDALE",	CONFIG_SYS_NAND_ALE_PIN,	0, PIO_PULLUP, PIO_PERIPH_A},
 		{"NANDCLE",	CONFIG_SYS_NAND_CLE_PIN,	0, PIO_PULLUP, PIO_PERIPH_A},
 		{"NANDCS", 	CONFIG_SYS_NAND_ENABLE_PIN,	1, PIO_PULLUP, PIO_OUTPUT},
-		{"NWP", AT91C_PIN_PD(10), 1, PIO_PULLUP, PIO_OUTPUT},
+		{"NWP", AT91C_PIN_PD(10), 0, PIO_DEFAULT, PIO_OUTPUT},
 		{(char *)0,	0, 0, PIO_DEFAULT, PIO_PERIPH_A},
 	};
 
-- 
1.7.9.5

